File: MCPForUnity/Editor/Tools/ManageLightmap.cs
using System;
using System.Collections.Generic;
using MCPForUnity.Editor.Helpers;
using Newtonsoft.Json.Linq;
using UnityEditor;
using UnityEngine;

namespace MCPForUnity.Editor.Tools
{
    [McpForUnityTool("manage_lightmap", AutoRegister = false)]
    public static class ManageLightmap
    {
        public static object HandleCommand(JObject @params)
        {
            var p = new ToolParams(@params);
            var actionResult = p.GetRequired("action");
            var actionError = actionResult.GetOrError(out string action);
            if (actionError != null) return actionError;

            action = action.ToLowerInvariant();

            try
            {
                switch (action)
                {
                    case "set_static": return SetLightmapStatic(@params, p);
                    case "configure_settings": return ConfigureSettings(@params, p);
                    case "bake": return Bake(@params, p);
                    case "clear": return Clear(@params, p);
                    case "set_lightmap_parameters": return SetLightmapParameters(@params, p);
                    case "get_info": return GetInfo(@params, p);
                    default:
                        return new ErrorResponse($"Unknown action: {action}. Valid: set_static, configure_settings, bake, clear, set_lightmap_parameters, get_info");
                }
            }
            catch (Exception ex)
            {
                return new ErrorResponse(ex.Message, new { stackTrace = ex.StackTrace });
            }
        }

        private static object SetLightmapStatic(JObject @params, ToolParams p)
        {
            var targetResult = p.GetRequired("target");
            var targetError = targetResult.GetOrError(out string target);
            if (targetError != null) return targetError;

            GameObject go = GameObject.Find(target);
            if (go == null) return new ErrorResponse($"GameObject '{target}' not found.");

            bool contribute = p.GetBool("contribute_gi", true);

            Undo.RecordObject(go, "Set Lightmap Static");
            StaticEditorFlags flags = GameObjectUtility.GetStaticEditorFlags(go);

            if (contribute)
                flags |= StaticEditorFlags.ContributeGI;
            else
                flags &= ~StaticEditorFlags.ContributeGI;

            GameObjectUtility.SetStaticEditorFlags(go, flags);

            if (p.GetBool("include_children", false))
            {
                foreach (Transform child in go.GetComponentsInChildren<Transform>())
                {
                    if (child.gameObject == go) continue;
                    Undo.RecordObject(child.gameObject, "Set Lightmap Static");
                    StaticEditorFlags cf = GameObjectUtility.GetStaticEditorFlags(child.gameObject);
                    if (contribute) cf |= StaticEditorFlags.ContributeGI;
                    else cf &= ~StaticEditorFlags.ContributeGI;
                    GameObjectUtility.SetStaticEditorFlags(child.gameObject, cf);
                }
            }

            // Set lightmap scale if specified
            float? scale = p.GetFloat("lightmap_scale");
            if (scale.HasValue)
            {
                Renderer renderer = go.GetComponent<Renderer>();
                if (renderer != null)
                {
                    SerializedObject so = new SerializedObject(renderer);
                    var scaleProp = so.FindProperty("m_ScaleInLightmap");
                    if (scaleProp != null)
                    {
                        scaleProp.floatValue = scale.Value;
                        so.ApplyModifiedProperties();
                    }
                }
            }

            return new SuccessResponse($"Set lightmap static on '{target}'", new { contributeGI = contribute });
        }

        private static object ConfigureSettings(JObject @params, ToolParams p)
        {
            // Lightmapping settings via LightmapEditorSettings
            SerializedObject so = new SerializedObject(
                UnityEngine.Object.FindObjectsByType<LightmapSettings>(FindObjectsSortMode.None).Length > 0
                    ? UnityEngine.Object.FindObjectsByType<LightmapSettings>(FindObjectsSortMode.None)[0]
                    : (UnityEngine.Object)Lightmapping.lightingSettings
            );

            // Use Lightmapping API directly
            float? indirectResolution = p.GetFloat("indirect_resolution");
            float? directResolution = p.GetFloat("direct_resolution");
            int? maxLightmapSize = p.GetInt("max_lightmap_size");
            int? bounces = p.GetInt("bounces");

            // Use LightingSettings if available
            var settings = Lightmapping.lightingSettings;
            if (settings != null)
            {
                Undo.RecordObject(settings, "Configure Lightmap Settings");

                if (indirectResolution.HasValue)
                    settings.indirectResolution = indirectResolution.Value;
                if (directResolution.HasValue)
                    settings.lightmapResolution = directResolution.Value;
                if (maxLightmapSize.HasValue)
                    settings.lightmapMaxSize = maxLightmapSize.Value;
                if (bounces.HasValue)
                    settings.maxBounces = bounces.Value;

                string lightmapper = p.Get("lightmapper");
                if (!string.IsNullOrEmpty(lightmapper))
                {
                    if (lightmapper.Equals("progressive_cpu", StringComparison.OrdinalIgnoreCase))
                        settings.lightmapper = LightingSettings.Lightmapper.ProgressiveCPU;
                    else if (lightmapper.Equals("progressive_gpu", StringComparison.OrdinalIgnoreCase))
                        settings.lightmapper = LightingSettings.Lightmapper.ProgressiveGPU;
                }

                if (p.Has("compress_lightmaps"))
                    settings.compressLightmaps = p.GetBool("compress_lightmaps", true);

                if (p.Has("ambient_occlusion"))
                    settings.ao = p.GetBool("ambient_occlusion", false);

                float? aoDistance = p.GetFloat("ao_max_distance");
                if (aoDistance.HasValue)
                    settings.aoMaxDistance = aoDistance.Value;

                EditorUtility.SetDirty(settings);
            }

            return new SuccessResponse("Configured lightmap settings");
        }

        private static object Bake(JObject @params, ToolParams p)
        {
            bool async = p.GetBool("async", true);

            if (async)
            {
                Lightmapping.BakeAsync();
                return new SuccessResponse("Lightmap bake started (async). Check progress in Lighting window.");
            }
            else
            {
                Lightmapping.Bake();
                return new SuccessResponse("Lightmap bake completed.");
            }
        }

        private static object Clear(JObject @params, ToolParams p)
        {
            Lightmapping.Clear();
            Lightmapping.ClearDiskCache();
            return new SuccessResponse("Cleared all lightmap data and disk cache.");
        }

        private static object SetLightmapParameters(JObject @params, ToolParams p)
        {
            var targetResult = p.GetRequired("target");
            var targetError = targetResult.GetOrError(out string target);
            if (targetError != null) return targetError;

            GameObject go = GameObject.Find(target);
            if (go == null) return new ErrorResponse($"GameObject '{target}' not found.");

            Renderer renderer = go.GetComponent<Renderer>();
            if (renderer == null) return new ErrorResponse($"No Renderer on '{target}'.");

            SerializedObject so = new SerializedObject(renderer);

            float? scale = p.GetFloat("lightmap_scale");
            if (scale.HasValue)
            {
                var scaleProp = so.FindProperty("m_ScaleInLightmap");
                if (scaleProp != null) scaleProp.floatValue = scale.Value;
            }

            string receiveGI = p.Get("receive_gi");
            if (!string.IsNullOrEmpty(receiveGI))
            {
                var giProp = so.FindProperty("m_ReceiveGI");
                if (giProp != null)
                {
                    if (receiveGI.Equals("lightmaps", StringComparison.OrdinalIgnoreCase))
                        giProp.intValue = 1;
                    else if (receiveGI.Equals("lightprobes", StringComparison.OrdinalIgnoreCase))
                        giProp.intValue = 2;
                }
            }

            so.ApplyModifiedProperties();

            return new SuccessResponse($"Set lightmap parameters on '{target}'");
        }

        private static object GetInfo(JObject @params, ToolParams p)
        {
            var settings = Lightmapping.lightingSettings;
            var info = new Dictionary<string, object>
            {
                { "isRunning", Lightmapping.isRunning },
                { "lightmapCount", LightmapSettings.lightmaps?.Length ?? 0 },
            };

            if (settings != null)
            {
                info["lightmapper"] = settings.lightmapper.ToString();
                info["indirectResolution"] = settings.indirectResolution;
                info["directResolution"] = settings.lightmapResolution;
                info["maxSize"] = settings.lightmapMaxSize;
                info["bounces"] = settings.maxBounces;
                info["compressLightmaps"] = settings.compressLightmaps;
                info["ao"] = settings.ao;
            }

            return new SuccessResponse("Lightmap info", info);
        }
    }
}

File: Server/src/services/tools/manage_lightmap.py
"""
Defines the manage_lightmap tool.
"""
from typing import Annotated, Any, Literal
from fastmcp import Context
from mcp.types import ToolAnnotations
from services.registry import mcp_for_unity_tool
from services.tools import get_unity_instance_from_context
from transport.unity_transport import send_with_unity_instance
from transport.legacy.unity_connection import async_send_command_with_retry

@mcp_for_unity_tool(
    description="Manages Unity lightmap baking: set GI static, configure lightmapper, bake (async/sync), clear, per-renderer params. Actions: set_static, configure_settings, bake, clear, set_lightmap_parameters, get_info.",
    annotations=ToolAnnotations(title="Manage Lightmap", destructiveHint=True),
)
async def manage_lightmap(
    ctx: Context,
    action: Annotated[Literal["set_static", "configure_settings", "bake", "clear", "set_lightmap_parameters", "get_info"], "Action"],
    target: Annotated[str, "Target GameObject"] | None = None,
    contribute_gi: Annotated[bool, "Mark as Contribute GI"] | None = None,
    include_children: Annotated[bool, "Apply to children"] | None = None,
    lightmap_scale: Annotated[float, "Scale in lightmap"] | None = None,
    indirect_resolution: Annotated[float, "Indirect resolution"] | None = None,
    direct_resolution: Annotated[float, "Direct resolution (lightmap resolution)"] | None = None,
    max_lightmap_size: Annotated[int, "Max lightmap size"] | None = None,
    bounces: Annotated[int, "Max bounces"] | None = None,
    lightmapper: Annotated[str, "Lightmapper: progressive_cpu, progressive_gpu"] | None = None,
    compress_lightmaps: Annotated[bool, "Compress lightmaps"] | None = None,
    ambient_occlusion: Annotated[bool, "Enable ambient occlusion"] | None = None,
    ao_max_distance: Annotated[float, "AO max distance"] | None = None,
    receive_gi: Annotated[str, "Receive GI mode: lightmaps, lightprobes"] | None = None,
) -> dict[str, Any]:
    unity_instance = get_unity_instance_from_context(ctx)
    params_dict = {k: v for k, v in {
        "action": action, "target": target, "contribute_gi": contribute_gi,
        "include_children": include_children, "lightmap_scale": lightmap_scale,
        "indirect_resolution": indirect_resolution, "direct_resolution": direct_resolution,
        "max_lightmap_size": max_lightmap_size, "bounces": bounces,
        "lightmapper": lightmapper, "compress_lightmaps": compress_lightmaps,
        "ambient_occlusion": ambient_occlusion, "ao_max_distance": ao_max_distance,
        "receive_gi": receive_gi,
    }.items() if v is not None}
    result = await send_with_unity_instance(async_send_command_with_retry, unity_instance, "manage_lightmap", params_dict)
    return result if isinstance(result, dict) else {"success": False, "message": str(result)}
